src/basic-multiclient.js

```diff
@@ -0,0 +1,103 @@
+const { EventEmitter } = require("events");
+const MarketObjectTypes = require("./enums");
+const winston = require("winston");
+
+class BasicMultiClient extends EventEmitter {
+  constructor(singleClientType) {
+    super();
+    this._clients = new Map();
+
+    this.singleClientType = singleClientType;
+  }
+
+  subscribeTicker(market) {
+    this._subscribe(market, this._clients, MarketObjectTypes.ticker, "subscribing to ticker");
+  }
+
+  unsubscribeTicker(market) {
+    if (this._clients.has(market.id)) {
+      this._clients.get(market.id).unsubscribeTicker(market);
+    }
+  }
+
+  subscribeTrades(market) {
+    this._subscribe(market, this._clients, MarketObjectTypes.trade, "subscribing to trades");
+  }
+
+  unsubscribeTrades(market) {
+    if (this._clients.has(market.id)) {
+      this._clients.get(market.id).unsubscribeTrades(market);
+    }
+  }
+
+  subscribeLevel2Updates(market) {
+    this._subscribe(
+      market,
+      this._clients,
+      MarketObjectTypes.l2update,
+      "subscribing to level 2 snapshots"
+    );
+  }
+
+  unsubscribeLevel2Updates(market) {
+    if (this._clients.has(market.id)) {
+      this._clients.get(market.id).unsubscribeLevel2Updates(market);
+    }
+  }
+
+  close(emitClosed = true) {
+    this._clients.forEach(c => {
+      c.close();
+    });
+
+    if (emitClosed) this.emit("closed");
+  }
+
+  _subscribe(market, map, marketObjectType, msg) {
```

Comment:
Perfect implementation of event propagation in this function.  

My only concern is that it looks like it's reusing the client on multiple "types" of subscriptions for the same market.  Is this allowed for Coinex?  I was thinking we would create a client for each type and each market.  Basically create a map for each type and create a new client for each type+market.  Thoughts?

Reply:
Oh, I thought that would be allowed. I thought it just wasn't allowed to reuse the client for the same type. 

I'm pretty sure I ran some tests with all 3 subscription types and got results in the command line. I should add a test around this to confirm it is working as expected and that it continues to deliver trades after i received other types of responses for example. I will double check that. If that doesn't work, then I will change it to do what you said.

Reply:
Cool sounds good.  And if we have a use case down the line where it has to be one call per sub, then we can easily modify `_subscribe` to accommodate both.  The code coverage around that conditions will ensure Coinex continues to function as expected.

Reply:
This was behaving as I thought it was. I did add a test around it, so like you said, if we have to refactor basic-multiclient.js to support other multi-markets in the future that don't behave like this, then we easily can.